### PR TITLE
Time Layer for measure the Store methods calls time

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Time Layer for measure the Store methods calls time and send the data to prometheus

### CODE REVIEW

File: app/server_app_adapters.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func (s *Server) RunOldAppInitalization() error {
 
 	if s.FakeApp().Srv.newStore == nil {
 		s.FakeApp().Srv.newStore = func() store.Store {
-			return store.NewLayeredStore(sqlstore.NewSqlSupplier(s.FakeApp().Config().SqlSettings, s.Metrics), s.Metrics, s.Cluster)
+			return store.NewTimerLayer(store.NewLayeredStore(sqlstore.NewSqlSupplier(s.FakeApp().Config().SqlSettings, s.Metrics), s.Metrics, s.Cluster), s.Metrics)
```
Comment (crspeller) [inline app/server_app_adapters.go]: Should this be optional? Not sure the performance implications of tracking metrics on every DB call.

Comment (jespino) [reply app/server_app_adapters.go]: The Prometheus performance implications is pretty low: here: https://prometheus.io/docs/introduction/faq/#what-is-the-performance-impact-of-instrumentation

Something like 12-17ns overhead per call, lets say 20ns, then 1million calls to the store adds around 20ms extra in total. I don't see the point on adding the extra setting.

Comment (jespino) [reply app/server_app_adapters.go]: The result of the benchmark is super weird, the layer on top of a mock store is giving more performance than accessing the mock directly

Here are my benchmark results:
```
goos: linux
goarch: amd64
pkg: github.com/mattermost/mattermost-server/store/bench
BenchmarkWithTimerLayerWithoutMetrics-4   	 1000000	     20339 ns/op
BenchmarkWithTimerLayer-4                 	 1000000	     21383 ns/op
BenchmarkWithoutTimerLayer-4              	 1000000	     34581 ns/op
PASS
ok  	github.com/mattermost/mattermost-server/store/bench	80.022s
```

and here the benchmark code:

```go
package bench

import (
    "testing"

    "github.com/mattermost/mattermost-server/enterprise/metrics"
    "github.com/mattermost/mattermost-server/plugin/plugintest/mock"
    "github.com/mattermost/mattermost-server/store"
    "github.com/mattermost/mattermost-server/store/storetest/mocks"
)

var storeWithTimer store.Store
var storeWithTimerWithoutMetrics store.Store
var storeWithoutTimer store.Store

func init() {
    metricsService := metrics.New(nil)
    storeMock := mocks.Store{}
    userStoreMock := mocks.UserStore{}
    userStoreMock.On("Get", mock.AnythingOfType("string")).Return(nil, nil)

    storeMock.On("User").Return(&userStoreMock)
    storeMock.On("Audit").Return(&mocks.AuditStore{})
    storeMock.On("Bot").Return(&mocks.BotStore{})
    storeMock.On("Channel").Return(&mocks.ChannelStore{})
    storeMock.On("ChannelMemberHistory").Return(&mocks.ChannelMemberHistoryStore{})
    storeMock.On("ClusterDiscovery").Return(&mocks.ClusterDiscoveryStore{})
    storeMock.On("Command").Return(&mocks.CommandStore{})
    storeMock.On("CommandWebhook").Return(&mocks.CommandWebhookStore{})
    storeMock.On("Compliance").Return(&mocks.ComplianceStore{})
    storeMock.On("Emoji").Return(&mocks.EmojiStore{})
    storeMock.On("FileInfo").Return(&mocks.FileInfoStore{})
    storeMock.On("Group").Return(&mocks.GroupStore{})
    storeMock.On("Job").Return(&mocks.JobStore{})
    storeMock.On("License").Return(&mocks.LicenseStore{})
    storeMock.On("LinkMetadata").Return(&mocks.LinkMetadataStore{})
    storeMock.On("OAuth").Return(&mocks.OAuthStore{})
    storeMock.On("Plugin").Return(&mocks.PluginStore{})
    storeMock.On("Post").Return(&mocks.PostStore{})
    storeMock.On("Preference").Return(&mocks.PreferenceStore{})
    storeMock.On("Reaction").Return(&mocks.ReactionStore{})
    storeMock.On("Role").Return(&mocks.RoleStore{})
    storeMock.On("Scheme").Return(&mocks.SchemeStore{})
    storeMock.On("Session").Return(&mocks.SessionStore{})
    storeMock.On("Status").Return(&mocks.StatusStore{})
    storeMock.On("System").Return(&mocks.SystemStore{})
    storeMock.On("Team").Return(&mocks.TeamStore{})
    storeMock.On("TermsOfService").Return(&mocks.TermsOfServiceStore{})
    storeMock.On("Token").Return(&mocks.TokenStore{})
    storeMock.On("User").Return(&mocks.UserStore{})
    storeMock.On("UserAccessToken").Return(&mocks.UserAccessTokenStore{})
    storeMock.On("UserTermsOfService").Return(&mocks.UserTermsOfServiceStore{})
    storeMock.On("Webhook").Return(&mocks.WebhookStore{})

    storeWithTimer = store.NewTimerLayer(&storeMock, metricsService)
    storeWithTimerWithoutMetrics = store.NewTimerLayer(&storeMock, nil)
    storeWithoutTimer = &storeMock
}

func BenchmarkWithTimerLayerWithoutMetrics(b *testing.B) {
    for n := 0; n < b.N; n++ {
        _, _ = storeWithTimerWithoutMetrics.User().Get("test")
    }
}

func BenchmarkWithTimerLayer(b *testing.B) {
    for n := 0; n < b.N; n++ {
        _, _ = storeWithTimer.User().Get("test")
    }
}

func BenchmarkWithoutTimerLayer(b *testing.B) {
    for n := 0; n < b.N; n++ {
        _, _ = storeWithoutTimer.User().Get("test")
    }
}
```

Comment (jespino) [reply app/server_app_adapters.go]: And I wrapped the SQL store tests with the TimerLayer, and everything is passing.

**Comment by amyblais** [comment]
@jespino Does this need a Jira ticket for QA testing?

**Comment by jwilander** [comment]
No this is just for developer testing/benchmarking

**Pull request merged**